### PR TITLE
chore: auto pick the only matching name on partial hits

### PR DESCRIPTION
**Pull request by hamidzr**
add a flag to skip this behavior

## Description
(this is a purely internal feature as it's nested under dev and "hidden" from users)
I find myself wanting this as I try out and test out changes otherwise I can never get the name right on the first try

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: harness/determined/cli/dev.py [hunk 1]
```diff
@@ -257,6 +259,16 @@ def auto_complete_binding(available_calls: List[str], fn_name: str) -> str:
     ]
     if not matches:
         raise errors.CliError(f"no such binding found: {fn_name}")
+    if not skip_partial_match and len(matches) == 1:
```
Comment (mackrorysd) [inline harness/determined/cli/dev.py:262]: Yes!! I thought this was a nuisance the other day, it'll be nice to automate getting past it. The name "skip_partial_match" isn't ideal IMHO - "skip_auto_picking" (or something else like that) which you use later seems clearer, and consistency would be good.

**Comment by hamidzr** [comment]
it's weird that our cli accepts a prefix of `--skip-au...` as the full option

File: harness/determined/cli/dev.py [hunk 2]
```diff
@@ -348,6 +358,11 @@ def call_bindings(args: argparse.Namespace) -> None:
                         "call a function from bindings",
                         [
                             cli.Arg("name", help="name of the function to call"),
+                            cli.Arg(
+                                "--skip-auto-picking",
```
Comment (MikhailKardash) [inline harness/determined/cli/dev.py:362]: The cli functionality here isn't to "skip" auto picking. It actually forces auto picking.
So I'd rename this arg to `auto-pick` or `skip-picking`

Comment (MikhailKardash) [reply harness/determined/cli/dev.py:362]: Though those names would also imply that you'd be force picking out of multiple hits.
Since you want it to match some regex, I think the best flag would be `--use-match`. This seems like a clearer description of what you want it to do. And if there are multiple matches, an extra step for disambiguation still makes sense.

File: harness/determined/cli/dev.py [hunk 3]
```diff
@@ -348,6 +358,11 @@ def call_bindings(args: argparse.Namespace) -> None:
...
+                                help="skip auto-picking of only matching name on partial hits",
```
Comment (MikhailKardash) [inline harness/determined/cli/dev.py:363]: Similar to the previous comment. I would reword this help text to say: "use matching function name if it is the only partial hit"

Comment (hamidzr) [reply harness/determined/cli/dev.py:363]: would that still keep the same default behavior? I'm proposing for the default behavior to be what it is right now where it accepts the first match if it's the only match.

Comment (MikhailKardash) [reply harness/determined/cli/dev.py:363]: Perhaps we have conflicting definitions of "auto-picking". Are you referring to auto-picking as bringing up the manual prompt? To me, if I see "auto-pick", I assume that means that the function itself is performing selection on behalf of the user somehow.

This function does one of 3 things:
1. call the binding function
2. if it finds one match in bindings, ask the user if that's what they're looking for an then continue
3. if it finds multiple matches, ask the user to disambiguate, then continue.

You want a flag that changes 2. Specifically, to skip asking the user if that's what they're looking for and instead continue with the regex'd binding function. It's essentially doing the same thing as `pip install -y`. Doing an auto-confirm. I want the flag to be some variation of "auto-confirm" and the description to also state something similar.

Comment (hamidzr) [reply harness/determined/cli/dev.py:363]: > You want a flag that changes 2. Specifically, to skip asking the user if that's what they're looking for and instead continue with the regex'd binding function. It's essentially doing the same thing as pip install -y

yes!
(before this pr points 2 and 3 are not differentiated at all)

> I want the flag to be some variation of "auto-confirm" and the description to also state something similar.

yeah auto confirm is a good way of putting it, that works for me. 

are you also asking that we don't default to auto confirming if there's only a single match? because my point with this pr is to make it the default but also add an option to get the old behavior if anyone wanted it

Comment (MikhailKardash) [reply harness/determined/cli/dev.py:363]: Yeah I'd rather have it not auto-confirm by default and require an `--auto-confirm` flag. It makes more sense to err on the side of caution and only add anoption for users that are absolutely sure that's what they want.

**Review by MikhailKardash — APPROVED** [review]
Stamping, since this is purely internal and "hidden" from users.
Internal users prefer auto-select to be on by default.

File: harness/determined/cli/dev.py (hunk 3 above)
Comment (hamidzr) [reply harness/determined/cli/dev.py]: yeah, that would generally be safer. It's hard to think through all the ways this could potentially lead the users to make the wrong API call 🤔 maybe we do go with the safer default and we'll keep using it with the flag for now and that'll give us a better idea. I could think of more complicated ways of trying to estimate the risks of auto confirming for the user, eg `get_*` calls can be assumed read-only and safe but we should probably keep it simple
